### ISSUE TITLE
Tree reorderable on tables

### ISSUE THREAD

**Issue by danharrin** [laravel-filament/filament#13324]

**Comment by johncarter-** [comment]
- Switch view toggle between standard table and tree editor.
 - Don't allow pagination.
 - If tree shows more than x rows show warning about performance?
 - Collapsible sub trees that can still be dragged with their children
 - Drag and drop trees between nested trees as well as being able to drag to top level.
 - Configurable attribute to act as title of draggable row/bar. E.g. title, slug, title and slug
 - Allow action / action group at end of draggable row/bar. To enable actions like delete, delete entire tree, create new child, etc.
 - Allow discard changes or save new order.

**Comment by rvzug** [comment]
I've come across this issue/suggestion. Could this work with the [Nested Set package by Lazychaser](https://github.com/lazychaser/laravel-nestedset)? I want to dive into making Filament compatible with this package (including reorder/drag&drop into other records/drag&drop between other records/drag&drop to parent level etc.), but if you are already planning something. I may be waiting...

I'm searching for a way to make this type of table: 
![nested-set](https://cdn.productived.net/media/pages/reviews/mylifeorganized/6a68635eff-1632301567/tasks-1-1-0x300.webp)

**Comment by GhostvOne** [comment]
- We should be able to access to the edit page of every items for relations manager

**Comment by binaryfire** [comment]
@danharrin Do you think there's a way of getting [this](https://github.com/filamentphp/filament/pull/8916) working as part of this update? Reordering tables is already a bit hard to follow without the usual Sortable.js visual feedback. It'll be worse with trees.

Something like this would be a big UX improvement: https://sortablejs.github.io/Sortable/#nested

**Comment by dmlogic** [comment]
Filament handling trees/nested sets would be amazing, and I'm happy to help implement any plan.

Our use case is Categories for which we use [laravel-nestedset](https://github.com/lazychaser/laravel-nestedset). We have two pain points:

Visually representing the tree when managing categories has been a challenge. With not a lot of available time we've ended up with:

* Viewing one level at a time with a breadcrumb generator in the infolist
* Relationship manager using a `children` `HasMany` relationship
* Custom actions to move up and down using the nested set methods

![Image](https://github.com/user-attachments/assets/c65595eb-de00-44b8-9b42-69f131ff68e0)

This works, but it's a poor representation of a tree.

Related to this is how to pick these these categories for relationships on other resources. Our current solution is horrible: a native Select list with a recursive options generator to prefix indents to give some kind of idea where you are. It totally falls apart when the list gets long or when you need to search. Feels like it needs a custom form control which allows you to explore the tree.

![Image](https://github.com/user-attachments/assets/795e7339-a26d-471b-b4a7-80829809478a)

**Comment by HelenaGwyn** [comment]
> Related to this is how to pick these these categories for relationships on other resources. Our current solution is horrible: a native Select list with a recursive options generator to prefix indents to give some kind of idea where you are. It totally falls apart when the list gets long or when you need to search. Feels like it needs a custom form control which allows you to explore the tree.

I've been looking into this as well. First I came accross this plugin https://github.com/15web/filament-tree which isn't ported to v4 yet. And I would like to stick to native Filament as much as possible.

For the selection of nested set I did this:
```
Forms\Components\Select::make('parent_id')
    ->label('Parent')
    ->searchable()
    ->relationship(name: 'parent', titleAttribute: 'name', modifyQueryUsing: fn (Builder $query) => $query->withDepth()->defaultOrder())
    ->getOptionLabelFromRecordUsing(function (Model $record) {
        $margin = $record->depth > 1 ? $record->depth * 15 : 0;
        $icon = $record->depth > 0 ? '<svg style="display: inline;margin-right: 10px;" class="fi-ta-text-item-icon h-5 w-5 text-gray-400 dark:text-gray-500" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" aria-hidden="true" data-slot="icon"><path stroke-linecap="round" stroke-linejoin="round" d="m16.49 12 3.75 3.75m0 0-3.75 3.75m3.75-3.75H3.74V4.499"></path></svg>' : '';
        $html = new HtmlString('<span style="margin-left: '.$margin.'px">'.$icon.$record->name.'</span>');
        return $html;
    })
```
It gives a nicer view of the tree.
<img width="423" height="385" alt="Image" src="https://github.com/user-attachments/assets/d311610d-2993-47e1-aa54-21d31d704ee2" />

For the table you can do something similar:
```
Tables\Columns\TextColumn::make('name')
    ->getStateUsing(function ($record) {
        $margin = $record->depth > 1 ? $record->depth * 15 : 0;
        $icon = $record->depth > 0 ? '<svg style="display: inline;margin-right: 10px;" class="fi-ta-text-item-icon h-5 w-5 text-gray-400 dark:text-gray-500" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" aria-hidden="true" data-slot="icon"><path stroke-linecap="round" stroke-linejoin="round" d="m16.49 12 3.75 3.75m0 0-3.75 3.75m3.75-3.75H3.74V4.499"></path></svg>' : '';
        return new HtmlString('<span style="margin-left: '.$margin.'px">'.$icon.$record->name.'</span>');
    })
```

<img width="440" height="537" alt="Image" src="https://github.com/user-attachments/assets/1777ded0-7088-4a01-a775-6b5fa46b1a65" />

Reordering still has to be done with seperate buttons on the records.

**Comment by danharrin** [comment]
There are a few plugins that do similar things to this now, and I think the implementation could end up being a little too opinionated for core, so I think it's probably best to close this.
